### ISSUE TITLE
RNN - Error: Shape (26, 50) must have rank at least 3

### ISSUE THREAD

**Issue by leonardomgomes** [silicon-valley-data-science/RNN-Tutorial#22]
Hello.
I'm a beginner in python and even more so in the tensorflow library.
I'm trying to run the code below but the tf.nn.dynamic_rnn method returns the error "Shape (26, 50) must have rank at least 3".
I'm holding the "input_data" parameter with 3D array but the error still appears.
I commented the snippet of code that used the "embedding_lookup" method because it only works with indexes (integers) and in my case, I'm working with real values ​​(audio conversion to MFCC).
Could someone please help me?
Thank you!

Note: I tried to use the "seq2seq.rnn_decoder" method but there were also errors.

import tensorflow as tf
from tensorflow.contrib import grid_rnn
from tensorflow.python.ops import rnn_cell
from tensorflow.contrib import legacy_seq2seq as seq2seq
from utils_audio import AudioLoader         #Biblioteca criada para converster arquivos de áudio em MFCC

#Identificando tipo de rede
model = 'lstm'
additional_cell_args = {}

#Escolha do tipo de modelo RNN
if model == 'rnn':
    cell_fn = rnn_cell.BasicRNNCell
elif model == 'gru':
    cell_fn = rnn_cell.GRUCell
elif model == 'lstm':
    cell_fn = rnn_cell.BasicLSTMCell
    additional_cell_args.update({'state_is_tuple': True})
elif model == 'gridlstm':
    cell_fn = grid_rnn.Grid2LSTMCell
    additional_cell_args.update({'use_peepholes': True, 'forget_bias': 1.0, 'state_is_tuple': True, 'output_is_tuple': True})
elif model == 'gridgru':
    cell_fn = grid_rnn.Grid2GRUCell
    additional_cell_args.update({'state_is_tuple': True, 'output_is_tuple': True})
else:
    raise Exception("model type not supported: {}".format(model))
    
#Definições do MFCC
num_cepstrum = 26 #Quantidade de cepstruns em cada unidade de tempo do áudio

#Definições dos parâmetros da rede RNN
rnn_hidden_units = int(128)  #size of RNN hidden state - quantidade de células da rnn.  Cada célula fica em uma replicação da RNN.
num_layers = 2  #number of layers in the RNN
batch_size = 50 #número de linhas de um batch (lote)
num_features = int(num_cepstrum) #26 - tamanho da sequencia de entrada na rede RNN
grad_clip = 5. #Grampeia gradientes neste valor
num_labels = 15

#Lendo arquivo de áudio e convertendo-o em objeto MFCC
audio_load = AudioLoader(batch_size, num_cepstrum)
#nomeAudio = "aahtm.wav"
nomeAudio = "abvwc.wav"
audio_load.wavToMFCC(nomeAudio, nomeAudio)    

#Instanciando a rede RNN
cell = cell_fn(rnn_hidden_units, **additional_cell_args) #método _init_ para instanciar uma célula do tipo de rede selecionada, definindo quantiade de unidades de estados da célula e argumentos adicionais
cell = rnn_cell.MultiRNNCell([cell] * num_layers, state_is_tuple=True) #cria a rede completa multiplicando a célula instanciada pela quantidade de camadas definida

#Instanciando objetos para os dados de entrada: input, label e estado inicial
input_data = tf.placeholder(tf.float32, [None, batch_size, num_features], name="input_data")#tf.placeholder - Insere um espaço reservado para um tensor que será sempre alimentado. O primeiro parâmetro é o tipo do elemento que será alimentado no Tensor.  O segundo parâmetro é o "shape" que é a configuração do Tensor que neste caso é uma matriz com as variáveis "batch_size" e "seq_length"
label = tf.placeholder(tf.string, [None, num_labels], name="label")
print("input_data: " + str(input_data))
initial_state = cell.zero_state(batch_size, tf.float32)
print("label: " + str(label))
print("initial_state: " + str(initial_state))

#Atribuindo batches do áudio MFCC para variáveis de entradas
'''
for i in range(audio_load.qtdBatchesMFCC):
    input_data = audio_load.arrayBatchesMFCC[i]
    label = "Maria"
'''
input_data = audio_load.arrayBatchesMFCC

with tf.variable_scope('rnnlm', reuse=tf.AUTO_REUSE) as rnnlm_scope:
    softmax_audio_w = tf.get_variable("softmax_audio_w", [rnn_hidden_units, num_labels])
    softmax_audio_b = tf.get_variable("softmax_audio_b", [num_labels, num_labels])
        
    with tf.device("/cpu:0"):        
        embedding_audio = tf.get_variable("embedding_audio", [num_labels, rnn_hidden_units])
        print("embedding_audio: " + str(embedding_audio))
        #lookup = tf.nn.embedding_lookup(embedding_audio, input_data)
        #print("embedding_lookup: " + str(lookup))
        #tf.nn.embedding_lookup(embedding_audio, input_data)
'''        inputs = tf.split(tf.nn.embedding_lookup(embedding_audio, input_data), num_or_size_splits=num_features, axis=1)
        inputs = [tf.squeeze(input_, [1]) for input_ in inputs]
        print("")
        print("inputs " + str(inputs))

        def loop(prev, _):
            prev = tf.nn.xw_plus_b(prev, softmax_audio_w, softmax_audio_b)#Computa os pesos e bias
            prev_symbol = tf.stop_gradient(tf.argmax(prev, 1))
            return tf.nn.embedding_lookup(embedding_audio, prev_symbol)    
    rnnlm_scope.reuse_variables()
'''

outputs, last_state = tf.nn.dynamic_rnn(cell, input_data, initial_state=initial_state, dtype=tf.float32)
#outputs, last_state = seq2seq.rnn_decoder(input_data, initial_state, cell, loop_function=None, scope=None)
